staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/envelope.go

```diff
@@ -142,13 +181,41 @@ func (t *envelopeTransformer) TransformToStorage(data []byte, context value.Cont
 	// Append the length of the encrypted DEK as the first 2 bytes.
 	b := cryptobyte.NewBuilder(nil)
 	b.AddUint16LengthPrefixed(func(b *cryptobyte.Builder) {
-		b.AddBytes([]byte(encKey))
+		b.AddBytes(dekCiphertext)
 	})
 	b.AddBytes(result)
 
 	return b.Bytes()
 }
 
+func (t *envelopeTransformer) rotateDEKIfRequired() ([]byte, value.Transformer, error) {
+	var (
+		err         error
+		transformer value.Transformer
+	)
+	t.mux.Lock()
+	defer t.mux.Unlock()
+
+	if !t.currentDEK.shouldRotate() {
+		if transformer, err = t.currentDEK.toTransformer(t.baseTransformerFunc); err == nil {
+			return t.currentDEK.ciphertext, transformer, nil
+		}
+
+		return nil, nil, err
+	}
+
+	if err := t.currentDEK.rotate(t.envelopeService, t.dekCacheTTL); err != nil {
+		return nil, nil, err
+	}
+
+	transformer, err = t.addTransformer(t.currentDEK.ciphertext, t.currentDEK.plaintext)
```

Comment:
can you reuse `t.currentDEK.toTransformer` here?
or maybe remove the `transformers` cache from `envelopeTransformer` entirely, since it's cached at the DEK level?

Reply:
PTAL, I think by inroducing newDEK I addressed most of of this.

We still need the cache for decryption - currentDEK only caches one DEK that is to be used by encryption.